File: lib/spells/effects/DemonSummon.cpp
/*
 * DemonSummon.cpp, part of VCMI engine
 *
 * Authors: listed in file AUTHORS in main folder
 *
 * License: GNU General Public License v2.0 or later
 * Full text of license available in license.txt file, in main folder
 *
 */
#include "StdInc.h"

#include "DemonSummon.h"
#include "Registry.h"
#include "../ISpellMechanics.h"
#include "../../battle/CBattleInfoCallback.h"
#include "../../battle/BattleInfo.h"
#include "../../battle/CUnitState.h"
#include "../../CStack.h"
#include "../../networkPacks/PacksForClientBattle.h"
#include "../../serializer/JsonSerializeFormat.h"

VCMI_LIB_NAMESPACE_BEGIN

namespace spells
{
namespace effects
{

int DemonSummon::raisedCreatureAmount(const Mechanics * m, const battle::Unit * unit) const
{
	if(!unit || unit->alive() || unit->isGhost())
		return 0;

	const auto *creatureType = creature.toEntity(m->creatures());

	int32_t deadCount         = unit->unitBaseAmount();
	int32_t deadTotalHealth   = unit->getTotalHealth();
	int32_t raisedMaxHealth   = creatureType->getMaxHealth();
	int32_t raisedTotalHealth = m->applySpellBonus(m->getEffectValue(), unit);

	// Can't raise stack with more HP than original stack
	int32_t maxAmountFromHealth     = deadTotalHealth / raisedMaxHealth;
	// Can't raise stack with more creatures than original stack
	int32_t maxAmountFromAmount     = deadCount;
	// Can't raise stack with more HP than our spellpower
	int32_t maxAmountFromSpellpower = raisedTotalHealth / raisedMaxHealth;

	int32_t finalAmount = std::min( { maxAmountFromHealth, maxAmountFromAmount, maxAmountFromSpellpower } );

	return finalAmount;
}

void DemonSummon::apply(ServerCallback * server, const Mechanics * m, const EffectTarget & target) const
{
	BattleUnitsChanged pack;
	pack.battleID = m->battle()->getBattle()->getBattleID();

	for(const Destination & dest : target)
	{
		const battle::Unit * targetStack = dest.unitValue;

		//we shall have all targets to be stacks
		if(!targetStack || targetStack->alive() || targetStack->isGhost())
		{
			server->complain("No corpse to demonize! Invalid effect target transformation.");
			continue;
		}

		auto hex = m->battle()->getAvailableHex(targetStack->creatureId(), m->casterSide, targetStack->getPosition().toInt());

		if(!hex.isValid())
		{
			server->complain("No place to put new summon!");
			break;
		}

		int32_t finalAmount = raisedCreatureAmount(m, targetStack);

		if(finalAmount < 1)
		{
			server->complain("Summoning didn't summon any!");
			continue;
		}

		battle::UnitInfo info;
		info.id       = m->battle()->battleNextUnitId();
		info.count    = finalAmount;
		info.type     = creature;
		info.side     = m->casterSide;
		info.position = dest.hexValue;
		info.summoned = !permanent;

		// add newly created creature
		pack.changedStacks.emplace_back(info.id, UnitChanges::EOperation::ADD);
		info.save(pack.changedStacks.back().data);

		// and remove corpse to prevent second raising or resurrection
		pack.changedStacks.emplace_back(targetStack->unitId(), UnitChanges::EOperation::REMOVE);
	}

	if(!pack.changedStacks.empty())
		server->apply(pack);
}

SpellEffectValue DemonSummon::getHealthChange(const Mechanics * m, const EffectTarget & spellTarget) const
{
	SpellEffectValue result;

	auto targets = m->getAffectedStacks(spellTarget);

	if(targets.empty())
		return result;

	auto unit = targets.front();
	if(unit)
	{
		result.unitsDelta = raisedCreatureAmount(m, unit);
		result.unitType = creature;
	}

	return result;
}

bool DemonSummon::isValidTarget(const Mechanics * m, const battle::Unit * unit) const
{
	if(!unit->isDead())
		return false;

	//check if alive unit blocks rising
	for(const BattleHex & hex : unit->getHexes())
	{
		auto blocking = m->battle()->battleGetUnitsIf([hex, unit](const battle::Unit * other)
		{
			return other->isValidTarget(false) && other->coversPos(hex) && other != unit;
		});

		if(!blocking.empty())
			return false;
	}

	if (unit->isGhost())
		return false;

	if (raisedCreatureAmount(m, unit) == 0)
		return false;

	return m->isReceptive(unit);
}

void DemonSummon::serializeJsonUnitEffect(JsonSerializeFormat & handler)
{
	handler.serializeId("id", creature, CreatureID());
	handler.serializeBool("permanent", permanent, false);
}

}
}

VCMI_LIB_NAMESPACE_END

File: scripts/demonSummon.lua
local Base = require("unitEffect")
local Script = setmetatable({}, {__index = Base})
Script.__index = Script

--- Returns the number of demons that can be raised from a dead unit.
function Script:raisedCreatureAmount(mechanics, unit)
	local creatureType = LIBRARY:getCreatureByName(self.id)

	local deadCount         = unit:getBaseAmount()
	local deadTotalHealth   = unit:getTotalHealth()
	local raisedMaxHealth   = creatureType:getMaxHealth()
	local raisedTotalHealth = mechanics:applySpellBonus(mechanics:getEffectValue(), unit)

	local maxFromHealth     = math.floor(deadTotalHealth   / raisedMaxHealth)
	local maxFromAmount     = deadCount
	local maxFromSpellpower = math.floor(raisedTotalHealth / raisedMaxHealth)

	return math.min(maxFromHealth, maxFromAmount, maxFromSpellpower)
end

--- A unit is a valid target if it is a dead, non-ghost corpse whose hexes are
--- not blocked by any alive unit, and our spellpower can raise at least one demon.
function Script:isValidTarget(mechanics, unit)
	if not unit:isDead() then return false end

	local hexes = unit:getHexes()
	for i = 1, hexes:size() do
		local hex = hexes:at(i)
		local blockers = mechanics:getBattle():getUnitsIf(function(other)
			return other:isValidTarget(false) and other:coversPos(hex)
		end)
		if #blockers > 0 then return false end
	end

	if unit:isGhost() then return false end

	if self:raisedCreatureAmount(mechanics, unit) == 0 then return false end

	return mechanics:isReceptive(unit)
end

--- Raise demons from each target corpse and remove the corpse.
function Script:apply(mechanics, server, target)
	local creatureType = LIBRARY:getCreatureByName(self.id)
	local battle       = mechanics:getBattle()

	for _, dest in ipairs(target) do
		local targetStack = dest.unit

		if targetStack == nil or not targetStack:isDead() or targetStack:isGhost() then
			print("DemonSummon: no valid corpse for demonization")
			break
		end

		local hex = battle:getAvailableHex(
			creatureType,
			mechanics:getCasterSide(),
			targetStack:getPosition()
		)

		if not hex:isValid() then
			print("DemonSummon: no available hex for summoned unit")
			break
		end

		local finalAmount = self:raisedCreatureAmount(mechanics, targetStack)

		if finalAmount < 1 then
			print("DemonSummon: raised zero creatures")
			break
		end

		server:addUnit(
			battle,
			{
				count    = finalAmount,
				type     = creatureType,
				side     = mechanics:getCasterSide(),
				position = hex,
				summoned = not self.permanent
			}
		)

		server:removeUnit(battle, targetStack)
	end
end

--- Returns the number of demons that would be raised for the hover preview.
function Script:getHealthChange(mechanics, spellTarget)
	local creatureType = LIBRARY:getCreatureByName(self.id)

	for _, dest in ipairs(spellTarget) do
		if dest.unit ~= nil then
			local amount = self:raisedCreatureAmount(mechanics, dest.unit)
			return {
				hpDelta    = 0,
				unitsDelta = amount,
				unitType   = creatureType
			}
		end
	end

	return { hpDelta = 0, unitsDelta = 0 }
end

return Script
